### PR TITLE
Unbounded length tests

### PR DESCRIPTION
**Pull request by pwalczysko**
# What this PR does

This PR adds 25 tests for long (apparently unbound) name lengths saving in the DB. The tests go over the 25 model objects modified in https://github.com/openmicroscopy/openmicroscopy/pull/4882. The tests make sure the names of the lenght 1MB are saved and retrieved to and from the DB of the 25 objects (the list of the objects is taken from the description of https://github.com/openmicroscopy/openmicroscopy/pull/4882).

Note that in cases where the fields in the DB are indexed, the test would fail with name lenghts of 1MB. Thus, the name lengths tested for was shortened in such cases to 2KB and negative tests were added for 1MB name lenghts for the relevant objects.
# Testing this PR

Make sure integration tests (java) are still passing. Look at the sensibility of the tests here.
# Related reading

https://github.com/openmicroscopy/openmicroscopy/pull/4882 and https://trello.com/c/ahSMgE9R/204-integration-tests-for-field-lengths?menu=filter&filter=@sebastienbesson

### CODE REVIEW

**Comment by pwalczysko** [comment]
cc @mtbc @joshmoore @jburel @sbesson

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 1]
```diff
@@ -0,0 +1,632 @@
+package integration;
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java:20]: need some OME copyright boilerplate probably:

[copyright.txt](https://github.com/openmicroscopy/openmicroscopy/files/536477/copyright.txt)

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import ome.api.JobHandle;
+
+public class ObjectPropertiesTest extends AbstractServerTest {
+
+    /**
+     * Creates a name with length as a string
+     * @param integer lenght of the name to create in bytes
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: length

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+     * @return the name
+     * @throws Exception unexpected
+     */
+    private String createName(int integer) throws Exception {
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: would a variable name like "length" be better? (alt+shift+R in Eclipse)

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+     */
+    @Test
+    public void testAnnotationNameAndNsSaving() throws Exception {
+
+        /* Tag annotation is used here as a good representative
+         * to test the annotations in general */
+        final TagAnnotation ann = new TagAnnotationI();
+
+        /* for annotation name sizes of >2KB the test fails */
+        /* createName() creates name with lenght in approx. bytes */
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: "length", maybe just search and replace :smiley:

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+        String name = createName(1000000);
+        ann.setName(omero.rtypes.rstring(name));
+        try {
+            iUpdate.saveAndReturnObject(ann);
+            Assert.fail("Hibernate operation: could not insert:..."
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: this is a strange failure message -- might be good to choose text that explains to a surprised tester what it means that the test failed?  (e.g., "expected not to be able to insert 1MB tag name into DB" or whatever)

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+            /* expected */
+        }
+
+        /* similarly, for annotation namespace sizes of >2KB the test fails */
+        String namespace = createName(1000000);
+        ann.setNs(omero.rtypes.rstring(namespace));
+
+        /* need to revert the name to 2KB size in order to be sure to test
+         * for namespace failure, not both name and namespace failure */
+        name = createName(2000);
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java:116]: aren't we supposed to do something with this name before the iUpdate?

**Comment by pwalczysko** [comment]
This should fix @mtbc 's comments.

**Comment by mtbc** [comment]
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration-java/lastCompletedBuild/testngreports/integration/ObjectPropertiesTest/

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+     */
+    @Test
+    public void testFolderNameSaving() throws Exception {
+        Folder dat = mmFactory.simpleFolder();
+        final String name = createName(1000000);
+        dat.setName(omero.rtypes.rstring(name));
+        Folder sent = (Folder) iUpdate.saveAndReturnObject(dat);
+        String savedName = sent.getName().getValue().toString();
+        long id = sent.getId().getValue();
+        final Folder retrievedDataset = (Folder) iQuery.get("Folder", id);
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: maybe we can rename `dat` and `retrievedDataset`

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+            /* expected */
+        }
+
+        /* similarly, for namespace display name sizes of >2KB the test fails */
+        String displayName = createName(1000000);
+        ns.setDisplayName(omero.rtypes.rstring(displayName));
+
+        /* need to revert the name to 2KB size in order to be sure to test
+         * for displayName failure, not both name and displayName failure */
+        name = createName(2000);
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java:318]: again need to actually set this name (done below but too late)

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+                      + "tag namespaces of size 1MB into the DB");
+        } catch (ServerError se) {
+            /* expected */
+        }
+
+        /* now set namespace with size 2KB (name was set to this size already above)
+         * and let the test pass */
+
+        namespace = createName(2000);
+        ann.setName(omero.rtypes.rstring(name));
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: already did this above

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+                      + "original file hashes of size 1MB into the DB");
+        } catch (ServerError se) {
+            /* expected */
+        }
+
+        /* now set hash with size 2KB (name is fine at 1MB as set above)
+         * and let the test pass */
+
+        hash = createName(2000);
+        oFile.setName(omero.rtypes.rstring(name));
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: already done above

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+        /* first prepare image and pixels to get valid rendering defs */
+        Image img = mmFactory.createImage();
+        Image sentI = (Image) iUpdate.saveAndReturnObject(img);
+        Pixels pixels = sentI.getPrimaryPixels();
+
+        /* now get the rendering defs back via a method and a query */
+        IRenderingSettingsPrx prx = factory.getRenderingSettingsService();
+        prx.setOriginalSettingsInSet(Pixels.class.getName(),
+                Arrays.asList(pixels.getId().getValue()));
+        final RenderingDef rDef = (RenderingDef) iQuery.findByQuery("select rdef from RenderingDef as rdef where rdef.pixels.id = :id",
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: `"FROM RenderingDef WHERE pixels.id = :id"` should suffice

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 12]
```diff
@@ -0,0 +1,652 @@
...
+        Image img = mmFactory.createImage();
+        Image sentI = (Image) iUpdate.saveAndReturnObject(img);
+        Pixels pixels = sentI.getPrimaryPixels();
+
+        /* now get the rendering defs back via a method and a query */
+        IRenderingSettingsPrx prx = factory.getRenderingSettingsService();
+        prx.setOriginalSettingsInSet(Pixels.class.getName(),
+                Arrays.asList(pixels.getId().getValue()));
+        final RenderingDef rDef = (RenderingDef) iQuery.findByQuery("select rdef from RenderingDef as rdef where rdef.pixels.id = :id", 
+                new ParametersI().addId(pixels.getId().getValue()));
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: out of curiosity is it okay without the `.getValue()`?

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 13]
```diff
@@ -0,0 +1,652 @@
...
+     */
+    @Test
+    public void testRoiNameSaving() throws Exception {
+
+        /* first prepare image to get valid roi */
+        Image img = mmFactory.createImageWithRoi();
+        Image sentI = (Image) iUpdate.saveAndReturnObject(img);
+        /* now get the roi back using a query */
+        final Roi roi = (Roi) iQuery.findByQuery("select roi from Roi as roi where roi.image.id = :id",
+                new ParametersI().addId(sentI.getId().getValue()));
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: same questions here as with RenderingDef test

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+    @Test
+    public void testWellExternalDescription() throws Exception {
+
+        /* create some plate which contains a valid well */
+        final Plate plate = mmFactory.createPlate(1, 1, 1, 1, false);
+        Plate sentP = (Plate) iUpdate.saveAndReturnObject(plate);
+
+        /* get the well back using a query */
+        final Well well = (Well) iQuery.findByQuery("select well from Well as well where well.plate.id = :id",
+                new ParametersI().addId(sentP.getId().getValue()));
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: same questions here as with rendering def test

**Comment by dominikl** [comment]
So annotation names are limited to <= 2kb but string fields of other objects are not?

**Comment by mtbc** [comment]
See yesterday's comment on https://trello.com/c/ahSMgE9R/204-integration-tests-for-field-lengths. Basically, it's the indexed columns that can't take a long string.

**Comment by dominikl** [comment]
Looks good to me. Nothing additional to add to @mtbc 's comments.

**Comment by jburel** [comment]
@pwalczysko: next steps?

**Comment by pwalczysko** [comment]
@jburel : Hoping to restart the work on this today and relist tomorrow. Was busy with the chown tests https://github.com/pwalczysko/openmicroscopy/commits/java-test-chown-all

**Comment by pwalczysko** [comment]
This should fix the comments of @mtbc , thank you.

File: components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java [hunk 15]
```diff
@@ -0,0 +1,650 @@
...
+                      + "tag namespaces of size 1MB into the DB");
+        } catch (ServerError se) {
+            /* expected */
+        }
+
+        /* now set namespace with size 2KB (name was set to this size already above)
+         * and let the test pass */
+
+        namespace = createName(2000);
+        ann.setNs(omero.rtypes.rstring(namespace));;
```
Comment (dominikl) [inline components/tools/OmeroJava/test/integration/ObjectPropertiesTest.java]: Double semicolon could be removed (sorry, missed that in the previous review).

**Comment by dominikl** [comment]
Looks good, all issues raised by @mtbc have been addressed. Tests passed, too. Just spotted some unnecessary double semicolons (bit nitpicky, doesn't cause any harm).

**Comment by pwalczysko** [comment]
This should fix @dominikl comment.

**Comment by dominikl** [comment]
Good to merge from my point of view. 👍

**Comment by jburel** [comment]
Thanks all

**Pull request merged**